components/encryption/src/manager/mod.rs

```diff
@@ -98,75 +101,85 @@ impl Dicts {
         }
     }
 
-    fn save_key_dict(&mut self, master_key: &dyn Backend) -> Result<()> {
+    fn save_key_dict(&self, master_key: &dyn Backend) -> Result<()> {
         let file = EncryptedFile::new(&self.base, KEY_DICT_NAME);
         if !master_key.is_secure() {
-            for value in self.key_dict.keys.values_mut() {
+            let mut key_dict = self.key_dict.write().unwrap();
```

Comment:
key_dict can use a similar logic to only hold a write_lock while doing IO, and take key_dict lock only to update in-memory data-structure. but depend on how it is implemented we may a separate write_lock.

Reply:
Yes. But the life time of this lock it outside of the IO. There is just to change `was_exposed `,